### PR TITLE
fix(lint): handle unused type parameters in overload signatures

### LINKED ISSUE #11214: 💅 Generic type that used only for type for variable considered as unused when function overload is used TS
### Environment information

If we define overload for function with generics - generic types considered as unused, but it is wrong

### Rule name

noUnusedVariables

### Playground link

https://biomejs.dev/playground/?lintRules=noUnusedVariables&tab=formatter&pane=Diagnostics&code=YQBzAHkAbgBjACAAZgB1AG4AYwB0AGkAbwBuACAAcgBlAHQAdQByAG4AQQBuAHkAKAApACAAewAKACAAIAByAGUAdAB1AHIAbgAgACIAIgAgAGEAcwAgAGEAbgB5AAoAfQAKAAoAZgB1AG4AYwB0AGkAbwBuACAAcwBvAG0AZQBGAG4APABNAHkARwBlAG4AZQByAGkAYwA%2BACgAKQAKAGYAdQBuAGMAdABpAG8AbgAgAHMAbwBtAGUARgBuADwATQB5AEcAZQBuAGUAcgBpAGMAPgAoACkAIAB7AAoAIAAgAGMAbwBuAHMAdAAgAGEAOgAgAE0AeQBHAGUAbgBlAHIAaQBjACAAPQAgAHIAZQB0AHUAcgBuAEEAbgB5ACgAKQAKACAAIABjAG8AbgBzAG8AbABlAC4AbABvAGcAKABhACkACgB9AAoAcwBvAG0AZQBGAG4AKAApAA%3D%3D&language=ts

### Expected result

It shouldn't throw error

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct

### PR DESCRIPTION
**Pull request by saberoueslati**
## Summary

Fixes #11214

`noUnusedVariables` incorrectly reports generic type parameters on overload signatures as unused when the implementation uses them. The analyzer now exempts only signatures with a matching body-bearing implementation while continuing to report unused parameters on abstract methods and interface members.

A patch changeset is included.

This PR was implemented with the help of Codex.

## Test Plan

- Added regression coverage for function and method overloads.
- Added coverage ensuring abstract and interface method type parameters remain diagnosed.
- `cargo test -p biome_js_analyze --test spec_tests correctness::no_unused_variables`
- `cargo fmt --all -- --check`

## Docs

N/A

### CODE REVIEW

**Review by MILLERMARRU — COMMENTED** [review]
Checked the TypeScript handbook on function overloads to make sure the premise here is right, not just plausible. Overload signatures never carry a body, only the implementation signature does, and the implementation itself isn't directly callable from outside, callers only ever see the overload signatures. That actually makes the old behavior structurally guaranteed to false-positive on every generic overload: a type parameter declared on a signature with no body has literally nowhere it could ever be "used" within that specific declaration, so checking each overload signature in isolation would flag every single one of them regardless of how the generic is actually used in the real implementation. Evaluating usage against the implementation instead, which is what this PR does, is the only approach that could ever produce a correct result here.

The edge cases in the new test fixture are the right ones to have thought through. Interface members and abstract methods staying reported makes sense since neither can ever gain a body-bearing counterpart the way a function/class overload can. The `MismatchedStaticness` case is a good one to have caught specifically, `static foo<T>(): void; foo(): void {}` isn't a valid overload pair in TypeScript, an instance method doesn't implement a static signature, so it's correct that `method_signature_has_implementation` checks static-ness equality before treating a same-named body-bearing method as the implementation. Missing that check would have silently exempted type parameters that genuinely have nowhere to be checked against.

`declare_function_has_implementation` relying on `signature.is_mergeable(&declaration)` rather than just a name match is also the safer choice, name collisions alone don't guarantee TypeScript would actually treat two declarations as one overload set.

File: crates/biome_js_analyze/src/lint/correctness/no_unused_variables.rs [hunk 1]
```diff
@@ -245,6 +249,156 @@ fn is_rest_spread_sibling(decl: &AnyJsBindingDeclaration) -> bool {
...
+/// Returns `true` if the type parameter belongs to a function or method
+/// signature (a declaration without a body, such as an overload signature)
+/// that has a body-bearing implementation of the same identity in scope.
+///
+/// TypeScript does not report unused type parameters on overload signatures.
+/// It checks the type parameters of the body-bearing implementation against
+/// references in that declaration's signature and body. A body-less method
+/// without an implementation, such as an abstract method, remains subject to
+/// this rule.
+/// See https://github.com/biomejs/biome/issues/11214
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/correctness/no_unused_variables.rs:261]: Rewrite this comment from scratch **using your own writing**.  What the hell is "body-bearing"? Plus three quarters of this comment is useless. Why do we care about typescript?

File: crates/biome_js_analyze/src/lint/correctness/no_unused_variables.rs [hunk 2]
```diff
@@ -245,6 +249,156 @@ fn is_rest_spread_sibling(decl: &AnyJsBindingDeclaration) -> bool {
...
+/// export per module, so its mere presence identifies it as the
+/// implementation of this signature.
+///
+/// Unlike [`declare_function_has_implementation`], this cannot go through the
+/// semantic model: a default export overload may be anonymous, in which case
+/// it binds no name to look up.
+fn declare_default_function_has_implementation(
+    signature: &TsDeclareFunctionExportDefaultDeclaration,
+) -> bool {
+    let Some(export_clause) = signature
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/correctness/no_unused_variables.rs:339]: Rewrite code based on our coding patterns. `Option<bool>` is an acceptable signature.

File: crates/biome_js_analyze/src/lint/correctness/no_unused_variables.rs [hunk 3]
```diff
@@ -245,6 +249,156 @@ fn is_rest_spread_sibling(decl: &AnyJsBindingDeclaration) -> bool {
...
+    )
+}
+
+/// Returns `true` if the name of `signature` is bound in the enclosing scope
+/// to a `function` declaration that TypeScript merges it with.
+///
+/// An overload signature always precedes its implementation, and a scope binds
+/// a name to the last declaration using it, so this lookup lands on the
+/// implementation whenever the overload set has one.
+fn declare_function_has_implementation(
```
Comment (ematipico) [inline crates/biome_js_analyze/src/lint/correctness/no_unused_variables.rs:303]: The comment is way more complex of the name itself. It doesn't add any value, it actually complicate things.

**Review by ematipico — CHANGES_REQUESTED** [review]
I am honestly rejecting this PR because it's full of garbage comments that don't mean anything.
